pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1473,13 +1508,19 @@ func (c *Cloud) DetachDisk(diskName string, instanceName string) (string, error)
 		return "", errors.New("no response from DetachVolume")
 	}
 
-	err = disk.waitForAttachmentStatus("detached")
+	attachment, err := disk.waitForAttachmentStatus("detached")
 	if err != nil {
 		return "", err
 	}
+	if attachment != nil {
+		// We expect it to be nil, it is (maybe) interesting if it is not
+		glog.V(2).Infof("waitForAttachmentStatus returned non-nil attachment with state=detached: %v", attachment)
+	}
 
 	if mountDevice != "" {
-		awsInstance.endAttaching(disk.awsID, mountDevice)
+		c.endAttaching(awsInstance, disk.awsID, mountDevice)
```

Comment:
When we call getMountDevice, we pass false, meaning we will never _add_ to the attaching list.  But if there was an in-process attach operation and we ended up detaching the disk, we believe that means that an in-flight attach operation has been cancelled.  So we call endAttaching, but we do not normally expect it to have been in the list.  Maybe it should now just be an error if there was a concurrent attach operation, or maybe the volume controller can now provide some sort of mutex guarantee?